_posts/2018-12-18-blog-a-more-complete-front-end-testing-strategy.md

```diff
@@ -0,0 +1,249 @@
+
+
+
+## What's wrong with the general front-end testing?
+
+When we are evaluating the integrity of a test strategy, we try to validate it with some of the following criteria:
+
+* Pass acceptance criteria
+* Catch bugs as early as possible
+* Faster running speed, especially at the top level of testing
+* Find bugs efficiently
+* Test writing costs and maintenance costs
+* Continuous refactoring risks
+
+I think a better test strategy should be as common as possible with the above items.
+
+In the general testing strategy mentioned above, E2E can override more AC conditions, but usually it runs less frequently; IT is run more frequently than E2E, but IT often includes an integrated app in almost the entire system, which is often more bloated in this case, and UT typically accounts for a larger proportion of such this strategy, although the logical coverage is good, but in a large refactoring usually UT will also change, of course, most of these UT is not too much of a problem.
+
+From the point of view of catching bugs as early as possible, does that mean that the general testing strategy can be better improved? Because the E2E test is most likely to run once in a relatively long period of time, rather than every PR or even every code commit, because the e2e run is usually slow and unstable, it is the most expensive automated test to run.
+
+Sometimes we have some integration tests that start a bloated integrated system, maybe it already includes a lot of mock, and you can keep running tests over and over again, but when there are more and more cases of integration testing, we can't even guarantee faster completion within each PR, and in a resource-constrained CI environment, it could be half an hour, or even longer.
+
+As a system becomes more complex, we need a complete testing strategy to tell us what cases failed in these tests, and to enable us to catch bugs more efficiently through test reports. Whether it is network instability, back-end server APIs exceptions, front-end domain modules exceptions or UI components exceptions, and so on, we can quickly catch these bugs. Obviously, the general testing strategy can provide limited help in locating bugs. For example, UT has succeeded, IT has failed, and E2E has failed too. It is difficult for us to analyze clearer information from such test reports.
+
+The cost of writing test code should be balanced with a continuous delivery development model. When the AC definition is clear, it's just that our test code should be able to overwrite the information described by AC. In theory, if all AC is fully implemented by E2E, this will also enable the verification of AC. But obviously, this brings with it a highly unbalanced test instability and inefficient operation.
```

Comment:
Again I think you should change the word 'overwrite'. To overwrite means to write over something and take over its place. Perhaps use 'align'? 'Our test code should align to the information described by AC' sounds about right.

Reply:
I agree.